sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/util/common/CounterNameAndMetadata.java

```diff
@@ -0,0 +1,155 @@
+ */
+package com.google.cloud.dataflow.sdk.util.common;
+
+import com.google.common.base.Strings;
+
+import javax.annotation.Nullable;
+
+/**
+ * A structured counter name.
+ *
+ * <p>It holds all information for generating structured name protos.
+ * It is also used as a wrapper during the migration to structured counters.
+ *
+ * <p>Before migration, the {@link CounterNameAndMetadata} will be converted to a flat name.
+ * During migration, the {@link CounterNameAndMetadata} will be mapped to a structured name proto.
+ */
+public class CounterNameAndMetadata {
+  /**
+   * Returns a CounterName with the given name.
+   */
+  public static CounterNameAndMetadata named(String name) {
+    return new CounterNameAndMetadata(name, "", "", "");
+  }
+
+  /**
+   * Returns a msecs CounterName.
+   */
+  public static CounterNameAndMetadata msecs(String name) {
+    return named(name + "-msecs");
+  }
+
+  /**
+   * Returns a CounterName identical to this, but with the given origin.
+   */
+  public CounterNameAndMetadata withOrigin(String origin) {
+    return new CounterNameAndMetadata(this.name, origin, this.stepName, this.prefix);
+  }
+
+  /**
+   * Returns a CounterName identical to this, but with the given original step name.
+   */
+  public CounterNameAndMetadata withOriginalStepName(String originalStepName) {
+    return new CounterNameAndMetadata(this.name, this.origin, originalStepName, this.prefix);
+  }
+
+  /**
+   * Returns a CounterName identical to this, but with the given prefix.
+   */
+  public CounterNameAndMetadata withPrefix(String prefix) {
+    return new CounterNameAndMetadata(this.name, this.origin, this.stepName, prefix);
+  }
+
+  /**
+   * Name of the counter.
+   *
+   * <p>For example, process-msecs, ElementCount.
+   */
+  private final String name;
+
+  /**
+   * Origin (namespace) of counter name.
+   *
+   * <p>For example, "user" for user-defined counters.
+   * It may be null or empty for counters defined by the Dataflow service or SDK.
+   */
+  @Nullable
+  private final String origin;
+
+  /**
+   * System defined step name.
+   *
+   * <p>For example, s1, s2.out.
+   * It may be null or empty when counters don't associate with step names.
+   */
+  @Nullable
+  private final String stepName;
+
+  /**
+   * Prefix of group of counters.
+   *
+   * <p>It is null or empty when counters don't have general prefixes.
+   */
+  @Nullable
+  private final String prefix;
+
+  /**
+   * Flat name is the equivalent unstructured name.
+   *
+   * <p>It is null before {@link #getFlatName()} is called.
+   */
+  @Nullable
+  private String flatName;
+
+  private CounterNameAndMetadata(String name, String origin, String stepName, String prefix) {
+    this.name = name;
+    this.origin = origin;
+    this.stepName = stepName;
+    this.prefix = prefix;
+    this.flatName = null;
+  }
+
+  /**
+   * Returns the flat name of a structured counter.
+   */
+  public synchronized String getFlatName() {
```

Comment:
If this is being called in multiple places, then we should only synchronize around the section that constructs the flatName (and modifies the field), if at all. Specifically, we shouldn't serialize all threads that may call `getFlatName` on a specific aggregator.